File: lab3/cs/StorageManager.cs
namespace Laba3_Semaphores_CSharp.classes;

public class StorageManager
{
  private Semaphore accessPermit;
  private Semaphore itemAvailable;
  private Semaphore spotsForNewItems;
  
  public readonly List<string> storage;
  
  public StorageManager(int storageSize, int amountItems)
  {
    accessPermit = new Semaphore(1, 1);
    itemAvailable = new Semaphore(0, storageSize);
    spotsForNewItems = new Semaphore(storageSize, storageSize);

    storage = new List<string>();
  }
    
  public String takeItem () {
    var item = storage.ElementAt(0);
    storage.RemoveAt(0);
    return item;
  }
	
  public void addItem(String item) {
    storage.Add(item);
  }
	
  public void getAccessPermit () {
    accessPermit.WaitOne();

  }
  public void dropAccessPermit () {
    accessPermit.Release();
  }
	
  public void getTakeItemPermit () {
    itemAvailable.WaitOne();
  }
  public void addNewTakeItemPermit () {
    itemAvailable.Release();
  }
	
  public void getAddItemPermit() {
    spotsForNewItems.WaitOne();
  }
	
  public void newAddItemPermit() {
    spotsForNewItems.Release();
  }
}

File: lab3/java/StorageManager.java
import java.util.ArrayList;
import java.util.concurrent.Semaphore;

public class StorageManager {
	private final Semaphore accessPermit = new Semaphore(1);
	private final Semaphore itemAvailable = new Semaphore(0);
	private final Semaphore spotsForNewItems;
	
	public ArrayList<String> storage = new ArrayList<>();
	
	public StorageManager(int storageSize) {
		spotsForNewItems = new Semaphore(storageSize);
	}
	
	public String takeItem () {
		var item = this.storage.get(0);
		this.storage.remove(0);
		return item;
	}
	
	public void addItem(String item) {
		this.storage.add(item);
	}
	
	public void getAccessPermit () {
		try {
			this.accessPermit.acquire();
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
	public void dropAccessPermit () {
		this.accessPermit.release();
	}
	
	public void getTakeItemPermit () {
		try {
			this.itemAvailable.acquire();
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
	public void addNewTakeItemPermit () {
		this.itemAvailable.release();
	}
	
	public void getAddItemPermit() {
		try {
			this.spotsForNewItems.acquire();
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
	
	public void newAddItemPermit() {
		this.spotsForNewItems.release();
	}
}
